### DISCUSSION TITLE
How do I set pass ChromeOptions using Driver(uc=True)?

### DISCUSSION THREAD

**Discussion by iamumairayub** [seleniumbase/SeleniumBase#2482] [Q&A] [answered]
How do I set pass `ChromeOptions` using `Driver(uc=True)`

```
from seleniumbase import Driver
from sbvirtualdisplay import Display
from selenium import webdriver
chrome_options = webdriver.ChromeOptions()

prefs = {"profile.managed_default_content_settings.images": 2}
chrome_options.add_experimental_option("prefs", prefs)

with Display(visible=0, size=(1440, 1880)):

    driver = Driver(
        uc=True, 
        browser='chrome',
        cap_string=chrome_options.to_capabilities()
    )
    driver.get('https://www.website.com/')
```

This code is to proof that these ChromeOptions are not being passed because website is still loading images.

**Comment by mdmintz** [comment, marked as answer]
Options are passed directly into the `Driver()` on initialization:

https://github.com/seleniumbase/SeleniumBase/blob/77ee5149a34a0bbca17c27a7e745d0754e3ad9e3/seleniumbase/plugins/driver_manager.py#L65

For example, there's already an option to block images with `block_images=True`:

https://github.com/seleniumbase/SeleniumBase/blob/77ee5149a34a0bbca17c27a7e745d0754e3ad9e3/seleniumbase/core/browser_launcher.py#L814

`cap_string` is only for Selenium Grid use with compatible grids. And the Selenium Grid isn't compatible with UC Mode, which means you can't use `cap_string` with UC Mode either. 

--------

Here's a full list of options available today:

```python
Driver(
    browser=None,  # Choose from "chrome", "edge", "firefox", or "safari".
    headless=None,  # The original headless mode for Chromium and Firefox.
    headless2=None,  # Chromium's new headless mode. (Has more features)
    headed=None,  # Run tests in headed/GUI mode on Linux, where not default.
    locale_code=None,  # Set the Language Locale Code for the web browser.
    protocol=None,  # The Selenium Grid protocol: "http" or "https".
    servername=None,  # The Selenium Grid server/IP used for tests.
    port=None,  # The Selenium Grid port used by the test server.
    proxy=None,  # Use proxy. Format: "SERVER:PORT" or "USER:PASS@SERVER:PORT".
    proxy_bypass_list=None,  # Skip proxy when using the listed domains.
    proxy_pac_url=None,  # Use PAC file. (Format: URL or USERNAME:PASSWORD@URL)
    multi_proxy=False,  # Allow multiple proxies with auth when multi-threaded.
    agent=None,  # Modify the web browser's User-Agent string.
    cap_file=None,  # The desired capabilities to use with a Selenium Grid.
    cap_string=None,  # The desired capabilities to use with a Selenium Grid.
    recorder_ext=None,  # Enables the SeleniumBase Recorder Chromium extension.
    disable_js=None,  # Disable JavaScript on websites. Pages might break!
    disable_csp=None,  # Disable the Content Security Policy of websites.
    enable_ws=None,  # Enable Web Security on Chromium-based browsers.
    disable_ws=None,  # Reverse of "enable_ws". (None and False are different)
    enable_sync=None,  # Enable "Chrome Sync" on websites.
    use_auto_ext=None,  # Use Chrome's automation extension.
    undetectable=None,  # Use undetected-chromedriver to evade bot-detection.
    uc_cdp_events=None,  # Capture CDP events in undetected-chromedriver mode.
    uc_subprocess=None,  # Use undetected-chromedriver as a subprocess.
    log_cdp_events=None,  # Capture {"performance": "ALL", "browser": "ALL"}
    no_sandbox=None,  # (DEPRECATED) - "--no-sandbox" is always used now.
    disable_gpu=None,  # (DEPRECATED) - GPU is disabled if not "swiftshader".
    incognito=None,  # Enable Chromium's Incognito mode.
    guest_mode=None,  # Enable Chromium's Guest mode.
    dark_mode=None,  # Enable Chromium's Dark mode.
    devtools=None,  # Open Chromium's DevTools when the browser opens.
    remote_debug=None,  # Enable Chrome's Debugger on "http://localhost:9222".
    enable_3d_apis=None,  # Enable WebGL and 3D APIs.
    swiftshader=None,  # Chrome: --use-gl=angle / --use-angle=swiftshader-webgl
    ad_block_on=None,  # Block some types of display ads from loading.
    host_resolver_rules=None,  # Set host-resolver-rules, comma-separated.
    block_images=None,  # Block images from loading during tests.
    do_not_track=None,  # Tell websites that you don't want to be tracked.
    chromium_arg=None,  # "ARG=N,ARG2" (Set Chromium args, ","-separated.)
    firefox_arg=None,  # "ARG=N,ARG2" (Set Firefox args, comma-separated.)
    firefox_pref=None,  # SET (Set Firefox PREFERENCE:VALUE set, ","-separated)
    user_data_dir=None,  # Set the Chrome user data directory to use.
    extension_zip=None,  # Load a Chrome Extension .zip|.crx, comma-separated.
    extension_dir=None,  # Load a Chrome Extension directory, comma-separated.
    disable_features=None,  # "F1,F2" (Disable Chrome features, ","-separated.)
    binary_location=None,  # Set path of the Chromium browser binary to use.
    driver_version=None,  # Set the chromedriver or uc_driver version to use.
    page_load_strategy=None,  # Set Chrome PLS to "normal", "eager", or "none".
    use_wire=None,  # Use selenium-wire's webdriver over selenium webdriver.
    external_pdf=None,  # Set Chrome "plugins.always_open_pdf_externally":True.
    is_mobile=None,  # Use the mobile device emulator while running tests.
    mobile=None,  # Shortcut / Duplicate of "is_mobile".
    d_width=None,  # Set device width
    d_height=None,  # Set device height
    d_p_r=None,  # Set device pixel ratio
    uc=None,  # Shortcut / Duplicate of "undetectable".
    undetected=None,  # Shortcut / Duplicate of "undetectable".
    uc_cdp=None,  # Shortcut / Duplicate of "uc_cdp_events".
    uc_sub=None,  # Shortcut / Duplicate of "uc_subprocess".
    log_cdp=None,  # Shortcut / Duplicate of "log_cdp_events".
    wire=None,  # Shortcut / Duplicate of "use_wire".
    pls=None,  # Shortcut / Duplicate of "page_load_strategy".
)
```

**Reply by joako233** [reply]
Hi! The argument of maximize= True does not work

**Reply by mdmintz** [reply]
Window maximization is done after the browser is launched:

```python
driver.maximize_window()
```

**Comment by joako233** [comment]
I have this error, thanks.
![michael](https://github.com/seleniumbase/SeleniumBase/assets/141453948/75e17836-ea8b-4346-87c1-6cef924ba9c9)

**Reply by mdmintz** [reply]
There's **no** `maximize=True` option for `Driver()` initialization.

You have to use `driver.maximize_window()` **after** the browser is launched.

**Reply by joako233** [reply]
Okay, thanks! I have another question.I want to know which browser arguments such as chromium_arg = "--lang-en" are added by default in SeleniumBase, since, as you told me, if I add the same argument twice, it can cause problems. Thanks again."

**Reply by mdmintz** [reply]
Search for them in [seleniumbase/core/browser_launcher.py](https://github.com/seleniumbase/SeleniumBase/blob/master/seleniumbase/core/browser_launcher.py).
There are also a few specific UC Mode options set in [`seleniumbase/undetected/__init__.py`](https://github.com/seleniumbase/SeleniumBase/blob/master/seleniumbase/undetected/__init__.py)

**Comment by telecran-telecrit** [comment]
`TypeError: Driver() got an unexpected keyword argument 'downloads_path'`

https://stackoverflow.com/questions/35331854/downloading-a-file-at-a-specified-location-through-python-and-selenium-using-chr/43789674#43789674

**Reply by mdmintz** [reply]
Duplicate of https://github.com/seleniumbase/SeleniumBase/issues/1557#issuecomment-1287636878

--------

SeleniumBase creates the ``downloaded_files`` folder in the working directory where ``pytest`` is invoked. Any click-initiated downloads will go there. It's also used for any special files needed. It also holds lock files to prevent issues with multi-threading. The folder resets at the start of every new ``pytest`` run so that past test runs don't interfere with new ones. The folder is hard-coded there to prevent issues. And since the folder is reset at the start of new ``pytest`` runs, you wouldn't want to use any other existing folders for it.

There are lots of built-in test methods that are specially made for that folder, such as:

```python
self.get_downloads_folder()

self.get_browser_downloads_folder()

self.get_path_of_downloaded_file(file)

self.is_downloaded_file_present(file)

self.delete_downloaded_file_if_present(file)
# Duplicates: self.delete_downloaded_file(file)

self.assert_downloaded_file(file)

self.get_downloaded_files(regex=None, browser=False)

self.get_data_from_downloaded_file(file, timeout=None, browser=False)

self.assert_data_in_downloaded_file(data, file, timeout=None, browser=False)
```

If you need files in a different folder, use Python ``os``, ``sys``, and ``shutil`` libraries for copying the files you need into a different folder after they get downloaded to the ``downloaded_files`` folder.

There's also a `pytest` command-line option to archive existing downloads instead of deleting them:
https://github.com/seleniumbase/SeleniumBase/blob/570910c6f6e079cddd00c23e826e202c6489f39e/README.md?plain=1#L673

(A copy of those downloads will go into a folder in the `archived_files/` folder at the end of the `pytest` run.)

**Reply by sebastian-hermix** [reply]
> Duplicate of [#1557 (comment)](https://github.com/seleniumbase/SeleniumBase/issues/1557#issuecomment-1287636878)
> 
> SeleniumBase creates the `downloaded_files` folder in the working directory where `pytest` is invoked. Any click-initiated downloads will go there. It's also used for any special files needed. It also holds lock files to prevent issues with multi-threading. The folder resets at the start of every new `pytest` run so that past test runs don't interfere with new ones. The folder is hard-coded there to prevent issues. And since the folder is reset at the start of new `pytest` runs, you wouldn't want to use any other existing folders for it.
> 
> There are lots of built-in test methods that are specially made for that folder, such as:
> 
> ```python
> self.get_downloads_folder()
> 
> self.get_browser_downloads_folder()
> 
> self.get_path_of_downloaded_file(file)
> 
> self.is_downloaded_file_present(file)
> 
> self.delete_downloaded_file_if_present(file)
> # Duplicates: self.delete_downloaded_file(file)
> 
> self.assert_downloaded_file(file)
> 
> self.get_downloaded_files(regex=None, browser=False)
> 
> self.get_data_from_downloaded_file(file, timeout=None, browser=False)
> 
> self.assert_data_in_downloaded_file(data, file, timeout=None, browser=False)
> ```
> 
> If you need files in a different folder, use Python `os`, `sys`, and `shutil` libraries for copying the files you need into a different folder after they get downloaded to the `downloaded_files` folder.
> 
> There's also a `pytest` command-line option to archive existing downloads instead of deleting them:
> 
> https://github.com/seleniumbase/SeleniumBase/blob/570910c6f6e079cddd00c23e826e202c6489f39e/README.md?plain=1#L673
> 
> (A copy of those downloads will go into a folder in the `archived_files/` folder at the end of the `pytest` run.)

Hello @mdmintz , could you tell me whether it’s possible to disable downloads to the downloaded_files directory, or if it must be cleared manually?

**Reply by mdmintz** [reply]
I think you can change it via CDP at runtime: https://github.com/mdmintz/MyCDP/blob/c84873e3aa0ca64c59da19f1a7c2cdab7f68556d/mycdp/browser.py#L398

There are examples of making mycdp calls from SB, eg: https://github.com/seleniumbase/SeleniumBase/blob/622998f9826ed80c4e6c910b8cf8c56e4f75cb90/examples/cdp_mode/raw_mobile_gitlab.py#L9-L11

but you'll need to use the `set_download_behavior` method. I think it exists in the async format: https://github.com/seleniumbase/SeleniumBase/blob/622998f9826ed80c4e6c910b8cf8c56e4f75cb90/seleniumbase/undetected/cdp_driver/tab.py#L1248-L1250

**Comment by estiven94ops** [comment]
Hi, is there any option for multiples downloads without navigator asking?
like prefs: "download.prompt_for_download": False,

**Reply by mdmintz** [reply]
You can already download multiple files without a prompt. For example:

```python
from seleniumbase import SB

with SB(test=True) as sb:
    sb.open("https://pypi.org/project/sbvirtualdisplay/#files")
    for i in range(3):
        sb.click('div.card a[href*=".tar.gz"]')
        sb.sleep(2)
```

**Reply by estiven94ops** [reply]
thank you!

**Comment by estiven94ops** [comment]
Hi, how can I open as a safe browser to download any files like .xml? using from seleniumbase import Driver

**Reply by mdmintz** [reply]
What do you mean by `open as a safe browser`?

If there's an option you need that isn't available in one of the other args, then add it to `chromium_arg`:
```python
chromium_arg=None,  # "ARG=N,ARG2" (Set Chromium args, ","-separated.)
```

**Reply by estiven94ops** [reply]
Hello, I am trying to download .xml files but Chrome automatically blocks dangerous downloads with a window asking the user if they want to download that file, is there an option with UC that allows the download of any file without consulting the user or showing the notification window?, I need to that automatically downloads any file; By clicking on the download (which is a .xml file), allow the download automatically

**Reply by mdmintz** [reply]
It shouldn't be because of this: https://github.com/seleniumbase/SeleniumBase/blob/b77a1c022bb4ebf8cc1fec4db92847a14f9bc818/seleniumbase/core/browser_launcher.py#L827
It's working correctly for me. I would need to see the exact message that appears for you to do more.

**Reply by estiven94ops** [reply]
ok...I open the driver only with these lines as a test, and download an .xml files:

from seleniumbase import Driver
from selenium.webdriver.support.ui import WebDriverWait

driver = Driver(uc=True)
driver.maximize_window()
wait = WebDriverWait(driver, 15)

driver.get("URL")
tbody_element = driver.find_element(By.ID,"frmPrincipal:tablaCompRecibidos_data")
onclick="mojarra.jsfcljs(document.getElementById('frmPrincipal'),{'frmPrincipal:tablaCompRecibidos:+0:lnkXml':'frmPrincipal:tablaCompRecibidos:0:lnkXml'},'');return false"
driver.execute_script(onclick)

I got a pop up window in chrome like this, because the driver is not open as "Enhanced Protection":
![c3a4c71c477162fc272e28c014ccabf2f6d271ae](https://github.com/seleniumbase/SeleniumBase/assets/71090662/2a0d14f1-8220-475f-a9f5-7714864ce01a)


Is any paramater I can use in "driver = Driver(uc=True)"?

**Reply by mdmintz** [reply]
SeleniumBase methods already include waits, so you don't need `WebDriverWait`.

Don't use `execute_script()` for clicking to download something.
Instead, use a regular click, eg. `driver.click(selector)` in SeleniumBase.

If that doesn't work, add `--unsafely-treat-insecure-origin-as-secure=*` into `chromium_arg` (replace `*` with the domain).
Reference: [https://stackoverflow.com/questions/78057740/chrome-122-how-to-allow-insecure-content-insecure-download-blocked/78065424#78065424](https://stackoverflow.com/a/78065424/7058266)

**Reply by estiven94ops** [reply]
ok, I tried using driver = Driver(uc=True, chromium_arg="--unsafely-treat-insecure-origin-as-secure=*") or Driver(uc=True, chromium_arg="--disable-features=InsecureDownloadWarnings " ), but didn´t work.
I also changed the download to this: driver.click("#frmPrincipal\\:tablaCompRecibidos\\:0\\:lnkXml"); In any case, the download continues to be blocked, with the same pop-up window appearing; another way to do it?

**Reply by mdmintz** [reply]
Did you replace `*` with the domain, as mentioned on StackOverflow? https://stackoverflow.com/a/78065424/7058266

This is not a SeleniumBase-specific issue, so if that StackOverflow solution didn't help, then there probably isn't more advice I can give you. Maybe it will work if you don't use UC Mode (if you don't need that).

**Reply by estiven94ops** [reply]
yes, I replaced it. I was trying to use UC: (https://github.com/seleniumbase/SeleniumBase/blob/master/help_docs/uc_mode.md) because the webSite has an  anti-bot service; is there any other way to use SeleniumBase and open driver in UC mode that allow me to disable thats pop up window with prefs or chrome.options?. In the "Undetectable Automation 2 video" you have 3 ways to open the chrome driver

**Reply by mdmintz** [reply]
You could try all those ways. You can even try `sb.download_file(file_url)` to download the file directly if you know the URL, but that's all I can think of, and that: https://stackoverflow.com/a/78065424/7058266

**Comment by Pranoy1c** [comment]
Is there a way to disable videos when using SeleniumBase with SB in CDP mode? Something like this:

```
from selenium.webdriver import ChromeOptions
chrome_options = ChromeOptions()
chrome_options.add_experimental_option(
    "prefs",
    {
        "profile.managed_default_content_settings.videos": 2  # 2 = Block videos
    }
)
```

**Reply by mdmintz** [reply]
Block the specific content that you want to block, like in https://github.com/seleniumbase/SeleniumBase/blob/master/examples/cdp_mode/raw_req_sb.py

**Comment by erfanedaraby** [comment]
hi
how can i tell seleniumbase not to close the browser after end of running the code?

in selenium we do this:
**from** selenium.webdriver.edge.options **import** Options
the_option=Options()
the_option.add_experimental_option("detach", True)
driver=webdriver.Edge(options=the_option)

**Reply by mdmintz** [reply]
Use a `breakpoint()` - https://github.com/seleniumbase/SeleniumBase/issues/1814#issuecomment-1478528980
